server/blob/mail-store/src/main/java/org/apache/james/blob/mail/MimeMessageStore.java

```diff
@@ -70,59 +73,29 @@ public Factory(BlobStore blobStore) {
     static class MimeMessageEncoder implements Store.Impl.Encoder<MimeMessage> {
         @Override
         public Stream<Pair<BlobType, Store.Impl.ValueToSave>> encode(MimeMessage message) {
+            Preconditions.checkNotNull(message);
             try {
-                byte[] messageAsArray = messageToArray(message);
-                int bodyStartOctet = computeBodyStartOctet(messageAsArray);
-                byte[] headerBytes = getHeaderBytes(messageAsArray, bodyStartOctet);
-                byte[] bodyBytes = getBodyBytes(messageAsArray, bodyStartOctet);
+                MessageParts messageParts = readMessage(message);
                 return Stream.of(
-                    Pair.of(HEADER_BLOB_TYPE, new Store.Impl.BytesToSave(headerBytes)),
-                    Pair.of(BODY_BLOB_TYPE, new Store.Impl.BytesToSave(bodyBytes)));
-            } catch (MessagingException | IOException e) {
+                    Pair.of(HEADER_BLOB_TYPE, new Store.Impl.InputStreamToSave(messageParts.getHeaderContent())),
+                    Pair.of(BODY_BLOB_TYPE, new Store.Impl.InputStreamToSave(messageParts.getBodyContent())));
+            } catch (IOException e) {
                 throw new RuntimeException(e);
             }
         }
-        
-        private static byte[] messageToArray(MimeMessage message) throws IOException, MessagingException {
-            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-            message.writeTo(byteArrayOutputStream);
-            return byteArrayOutputStream.toByteArray();
-        }
 
-        private static byte[] getHeaderBytes(byte[] messageContentAsArray, int bodyStartOctet) {
-            ByteBuffer headerContent = ByteBuffer.wrap(messageContentAsArray, 0, bodyStartOctet);
-            byte[] headerBytes = new byte[bodyStartOctet];
-            headerContent.get(headerBytes);
-            return headerBytes;
-        }
+        private static MessageParts readMessage(MimeMessage message) throws IOException {
+            PipedOutputStream outputStream = new PipedOutputStream();
```

Comment:
I'm not sure we need "piping" here (which is blocking one thread here) while we could write an outputstream version directly...